File: packages/client-search/src/types/Rule.ts
import { SearchOptions } from '.';

export type Rule = {
  /**
   * Unique identifier for the rule (format: [A-Za-z0-9_-]+).
   */
  readonly objectID: string;

  /**
   * Condition of the rule, expressed using the following variables: pattern, anchoring, context.
   *
   * @deprecated This parameter is deprecated in favor of `conditions`.
   */
  readonly condition?: Condition;

  /**
   * Conditions of the rule, expressed using the following variables: pattern, anchoring, context.
   */
  readonly conditions?: readonly Condition[];

  /**
   * Consequence of the rule. At least one of the following object must be used: params, promote, hide, userData.
   */
  readonly consequence?: Consequence;

  /**
   * This field is intended for rule management purposes, in particular to ease searching for rules and presenting them to human readers. It is not interpreted by the API.
   */
  readonly description?: string;

  /**
   * Whether the rule is enabled. Disabled rules remain in the index, but are not applied at query time.
   */
  readonly enabled?: boolean;

  /**
   * By default, rules are permanently valid. When validity periods are specified, the rule applies only during those periods; it is ignored the rest of the time.
   * The list must not be empty.
   */
  readonly validity?: readonly TimeRange[];
};

export type AutomaticFacetFilter = {
  /**
   * Attribute to filter on. This must match a facet placeholder in the rule’s pattern.
   */
  readonly facet: string;

  /**
   * Whether the filter is disjunctive (true) or conjunctive (false).
   */
  readonly disjunctive?: boolean;

  /**
   * Score for the filter. Typically used for optional or disjunctive filters.
   */
  readonly score?: number;
};

export type ConsequenceQuery = {
  /**
   * List of removes.
   */
  readonly remove?: readonly string[];

  /**
   * List of edits.
   */
  readonly edits?: ReadonlyArray<{
    /**
     * Type of edit.
     */
    readonly type?: 'remove' | 'replace';

    /**
     * Text or patterns to remove from the query string.
     */
    readonly delete?: string;

    /**
     * Text that should be inserted in place of the removed text inside the query string.
     */
    readonly insert?: string;
  }>;
};

export type ConsequencePromote =
  | {
      /**
       * Unique identifier of the object to promote.
       */
      readonly objectID: string;

      /**
       * Promoted rank for the object (zero-based).
       */
      readonly position: number;
    }
  | {
      /**
       * List of unique identifiers for the objects to promote.
       */
      readonly objectIDs: readonly string[];

      /**
       * Promoted start rank for the objects (zero-based).
       */
      readonly position: number;
    };

export type ConsequenceParams = {
  /**
   * When providing a string, it replaces the entire query string.
   * When providing an object, it describes incremental edits to be made to the query string (but you can’t do both).
   */
  readonly query?: ConsequenceQuery | string;

  /**
   * Names of facets to which automatic filtering must be applied; they must match the facet name of a facet value placeholder in the query pattern.
   */
  readonly automaticFacetFilters?: readonly AutomaticFacetFilter[] | readonly string[];

  /**
   * Same syntax as automaticFacetFilters, but the engine treats the filters as optional.
   * Behaves like optionalFilters.
   */
  readonly automaticOptionalFacetFilters?: readonly AutomaticFacetFilter[];
};

export type Condition = {
  /**
   * Query patterns are expressed as a string with a specific syntax. A pattern is a sequence of tokens.
   */
  readonly pattern?: string;

  /**
   * Apply this rule only when the filter matches.
   */
  readonly filters?: string;

  /**
   * is | startsWith | endsWith | contains: Whether the pattern must match the beginning or the end of the query string, or both, or none.
   */
  readonly anchoring?: 'is' | 'startsWith' | 'endsWith' | 'contains';

  /**
   * Rule context (format: [A-Za-z0-9_-]+). When specified, the rule is contextual and applies only when the same context is specified at query time (using the ruleContexts parameter).
   * When absent, the rule is generic and always applies (provided that its other conditions are met, of course).
   */
  readonly context?: string;

  /**
   * If set to true, alternatives make the rule to trigger on synonyms, typos and plurals.
   * Note that setting ignorePlurals to false overrides this parameter.
   */
  readonly alternatives?: boolean;
};

export type Consequence = {
  /**
   * Additional search parameters. Any valid search parameter is allowed.
   */
  readonly params?: ConsequenceParams & Pick<SearchOptions, Exclude<keyof SearchOptions, 'query'>>;

  /**
   * Objects to promote as hits.
   */
  readonly promote?: readonly ConsequencePromote[];

  /**
   * Objects to hide from hits.
   */
  readonly hide?: ReadonlyArray<{ readonly objectID: string }>;

  /**
   * Whether the Query Rule should promote or not promoted items.
   */
  readonly filterPromotes?: boolean;

  /**
   * Custom JSON object that will be appended to the userData array in the response.
   * This object is not interpreted by the API. It is limited to 1kB of minified JSON.
   */
  readonly userData?: any;
};

export type TimeRange = {
  /**
   * DateTime with UTC offset for Serialization/Deserialization in unix timespam.
   */
  readonly from: number;

  /**
   * DateTime with UTC offset for Serialization/Deserialization in unix timespam.
   */
  readonly until: number;
};

File: clients/algoliasearch-client-java-2/algoliasearch-core/com/algolia/model/Rule.java
package com.algolia.model;

import com.google.gson.annotations.SerializedName;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/** Rule object. */
@ApiModel(description = "Rule object.")
public class Rule {

  public static final String SERIALIZED_NAME_OBJECT_I_D = "objectID";

  @SerializedName(SERIALIZED_NAME_OBJECT_I_D)
  private String objectID;

  public static final String SERIALIZED_NAME_CONDITIONS = "conditions";

  @SerializedName(SERIALIZED_NAME_CONDITIONS)
  private List<Condition> conditions = null;

  public static final String SERIALIZED_NAME_CONSEQUENCE = "consequence";

  @SerializedName(SERIALIZED_NAME_CONSEQUENCE)
  private Consequence consequence;

  public static final String SERIALIZED_NAME_DESCRIPTION = "description";

  @SerializedName(SERIALIZED_NAME_DESCRIPTION)
  private String description;

  public static final String SERIALIZED_NAME_ENABLED = "enabled";

  @SerializedName(SERIALIZED_NAME_ENABLED)
  private Boolean enabled = true;

  public static final String SERIALIZED_NAME_VALIDITY = "validity";

  @SerializedName(SERIALIZED_NAME_VALIDITY)
  private List<TimeRange> validity = null;

  public Rule objectID(String objectID) {
    this.objectID = objectID;
    return this;
  }

  /**
   * Unique identifier of the object.
   *
   * @return objectID
   */
  @javax.annotation.Nonnull
  @ApiModelProperty(required = true, value = "Unique identifier of the object.")
  public String getObjectID() {
    return objectID;
  }

  public void setObjectID(String objectID) {
    this.objectID = objectID;
  }

  public Rule conditions(List<Condition> conditions) {
    this.conditions = conditions;
    return this;
  }

  public Rule addConditionsItem(Condition conditionsItem) {
    if (this.conditions == null) {
      this.conditions = new ArrayList<>();
    }
    this.conditions.add(conditionsItem);
    return this;
  }

  /**
   * A list of conditions that should apply to activate a Rule. You can use up to 25 conditions per
   * Rule.
   *
   * @return conditions
   */
  @javax.annotation.Nullable
  @ApiModelProperty(
    value = "A list of conditions that should apply to activate a Rule. You can use up to 25" +
    " conditions per Rule."
  )
  public List<Condition> getConditions() {
    return conditions;
  }

  public void setConditions(List<Condition> conditions) {
    this.conditions = conditions;
  }

  public Rule consequence(Consequence consequence) {
    this.consequence = consequence;
    return this;
  }

  /**
   * Get consequence
   *
   * @return consequence
   */
  @javax.annotation.Nonnull
  @ApiModelProperty(required = true, value = "")
  public Consequence getConsequence() {
    return consequence;
  }

  public void setConsequence(Consequence consequence) {
    this.consequence = consequence;
  }

  public Rule description(String description) {
    this.description = description;
    return this;
  }

  /**
   * This field is intended for Rule management purposes, in particular to ease searching for Rules
   * and presenting them to human readers. It's not interpreted by the API.
   *
   * @return description
   */
  @javax.annotation.Nullable
  @ApiModelProperty(
    value = "This field is intended for Rule management purposes, in particular to ease searching for" +
    " Rules and presenting them to human readers. It's not interpreted by the API."
  )
  public String getDescription() {
    return description;
  }

  public void setDescription(String description) {
    this.description = description;
  }

  public Rule enabled(Boolean enabled) {
    this.enabled = enabled;
    return this;
  }

  /**
   * Whether the Rule is enabled. Disabled Rules remain in the index, but aren't applied at query
   * time.
   *
   * @return enabled
   */
  @javax.annotation.Nullable
  @ApiModelProperty(
    value = "Whether the Rule is enabled. Disabled Rules remain in the index, but aren't applied at" +
    " query time."
  )
  public Boolean getEnabled() {
    return enabled;
  }

  public void setEnabled(Boolean enabled) {
    this.enabled = enabled;
  }

  public Rule validity(List<TimeRange> validity) {
    this.validity = validity;
    return this;
  }

  public Rule addValidityItem(TimeRange validityItem) {
    if (this.validity == null) {
      this.validity = new ArrayList<>();
    }
    this.validity.add(validityItem);
    return this;
  }

  /**
   * By default, Rules are permanently valid. When validity periods are specified, the Rule applies
   * only during those periods; it's ignored the rest of the time. The list must not be empty.
   *
   * @return validity
   */
  @javax.annotation.Nullable
  @ApiModelProperty(
    value = "By default, Rules are permanently valid. When validity periods are specified, the Rule" +
    " applies only during those periods; it's ignored the rest of the time. The list" +
    " must not be empty."
  )
  public List<TimeRange> getValidity() {
    return validity;
  }

  public void setValidity(List<TimeRange> validity) {
    this.validity = validity;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Rule rule = (Rule) o;
    return (
      Objects.equals(this.objectID, rule.objectID) &&
      Objects.equals(this.conditions, rule.conditions) &&
      Objects.equals(this.consequence, rule.consequence) &&
      Objects.equals(this.description, rule.description) &&
      Objects.equals(this.enabled, rule.enabled) &&
      Objects.equals(this.validity, rule.validity)
    );
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      objectID,
      conditions,
      consequence,
      description,
      enabled,
      validity
    );
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class Rule {\n");
    sb.append("    objectID: ").append(toIndentedString(objectID)).append("\n");
    sb
      .append("    conditions: ")
      .append(toIndentedString(conditions))
      .append("\n");
    sb
      .append("    consequence: ")
      .append(toIndentedString(consequence))
      .append("\n");
    sb
      .append("    description: ")
      .append(toIndentedString(description))
      .append("\n");
    sb.append("    enabled: ").append(toIndentedString(enabled)).append("\n");
    sb.append("    validity: ").append(toIndentedString(validity)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces (except the first line).
   */
  private String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }
}
